pkg/filters/opafilter/opafilter.go

```diff
@@ -0,0 +1,241 @@
+		logger.Errorf("create PrepareForEval rego query error: %s", err)
+	} else {
+		o.regoQuery = &query
+	}
+}
+
+func (o *OPAFilter) Inherit(previousGeneration filters.Filter) {
+	o.Init()
+	previousGeneration.Close()
+}
+
+func (o *OPAFilter) Handle(ctx *context.Context) (result string) {
+	req := ctx.GetInputRequest().(*httpprot.Request)
+	var rw *httpprot.Response
+	if rw, _ = ctx.GetOutputResponse().(*httpprot.Response); rw == nil {
+		rw, _ = httpprot.NewResponse(nil)
+		ctx.SetOutputResponse(rw)
+	}
+	return o.evalRequest(req, rw)
+}
+
+func (o *OPAFilter) Status() interface{} {
+	return nil
+}
+
+func (o *OPAFilter) Close() {
+}
+
+func (o *OPAFilter) evalRequest(r *httpprot.Request, w *httpprot.Response) string {
+	headers := map[string]string{}
+
+	for key, value := range r.HTTPHeader() {
+		if len(value) > 0 && stringtool.StrInSlice(key, o.includedHeadersParsed) {
+			headers[key] = strings.Join(value, ", ")
+		}
+	}
+
+	var body string
+	if o.spec.ReadBody {
+		buf, _ := io.ReadAll(r.Body)
+		body = string(buf)
+		// Put the body back in the request
+		r.Body = io.NopCloser(bytes.NewBuffer(buf))
+	}
+	pathParts := strings.Split(strings.Trim(r.Path(), "/"), "/")
+	input := map[string]any{
+		"request": map[string]any{
+			"method":     r.Std().Method,
+			"path":       r.Std().URL.Path,
+			"path_parts": pathParts,
+			"raw_query":  r.Std().URL.RawQuery,
+			"query":      map[string][]string(r.Std().URL.Query()),
+			"headers":    headers,
+			"scheme":     r.Scheme(),
+			"realIP":     r.RealIP(),
+			"body":       body,
+		},
+	}
+	results, err := o.regoQuery.Eval(r.Context(), rego.EvalInput(input))
+	if err != nil {
+		return o.opaError(w, err)
+	}
+
+	if len(results) == 0 {
+		return o.opaError(w, errOpaNoResult)
+	}
+	bp, err := o.handleRegoResult(w, results[0].Bindings["result"])
+	if err != nil {
+		return o.opaError(w, err)
+	}
+	if *bp {
+		return ""
+	}
+	return resultFiltered
+}
+
+func boolP(val bool) *bool {
+	return &val
+}
+
+func (o *OPAFilter) handleRegoResult(w *httpprot.Response, result any) (*bool, error) {
+	if allowed, ok := result.(bool); ok {
+		if !allowed {
+			w.SetStatusCode(o.spec.DefaultStatus)
+			return boolP(false), nil
+		}
+		return boolP(true), nil
+	}
+	if _, ok := result.(map[string]any); !ok {
+		return nil, errOpaInvalidResultType
+	}
+	marshaled, err := json.Marshal(result)
+	if err != nil {
+		return nil, errOpaInvalidResultType
+	}
+	regoResult := RegoResult{
+		// By default, a non-allowed request with return a 403 response.
+		StatusCode:        o.spec.DefaultStatus,
+		AdditionalHeaders: make(map[string]string),
+	}
+	if err = json.Unmarshal(marshaled, &regoResult); err != nil {
+		return nil, errOpaInvalidResultType
+	}
+	// Set the headers on the ongoing request (overriding as necessary)
+	for key, value := range regoResult.AdditionalHeaders {
+		w.Header().Set(key, value)
+	}
+	if !regoResult.Allow {
+		w.SetStatusCode(regoResult.StatusCode)
+	}
+	return boolP(regoResult.Allow), nil
```

Comment:
this is ok for now, but I think we can improve the performance by using type assertion to check the result directly, as `json.Unmarshal` & `json.Marshal` are slow.

Reply:
maybe we can use variables like results[0].Bindings["status_code"] and results[0].Bindings["additional_headers"] directly to avoid data convertion.